tests/components/vera/common.py

```diff
@@ -0,0 +1,856 @@
+
+    if expected_state is not None:
+        assert state.state == expected_state, (
+            "Expected '%s' (type %s) but was '%s' (type %s) for entity '%s'"
+            % (
+                expected_state,
+                type(expected_state),
+                state.state,
+                type(state.state),
+                entity_id,
+            )
+        )
+
+    if expected_brightness is not None:
+        assert_state_attribute(state, "brightness", expected_brightness)
+
+    if expected_hs_color is not None:
+        assert_state_attribute(state, "hs_color", expected_hs_color)
+
+    if expected_hvac_mode is not None:
+        assert_state_attribute(state, "hvac_mode", expected_hvac_mode)
+
+    if expected_fan_mode is not None:
+        assert_state_attribute(state, "fan_mode", expected_fan_mode)
+
+    if expected_temperature is not None:
+        assert_state_attribute(state, "temperature", expected_temperature)
+
+    if expected_current_temperature is not None:
+        assert_state_attribute(
+            state, "current_temperature", expected_current_temperature
+        )
+
+    if expected_min_temp is not None:
+        assert_state_attribute(state, "min_temp", expected_min_temp)
+
+    if expected_max_temp is not None:
+        assert_state_attribute(state, "max_temp", expected_max_temp)
+
+    if expected_current_position is not None:
+        assert_state_attribute(state, "current_position", expected_current_position)
+
+
+def assert_state_attribute(state: State, attribute: str, expected: Any) -> None:
+    """Assert a state's attribute has a specific value."""
+    assert attribute in state.attributes, "Attribute '%s' was not found." % attribute
+
+    value = state.attributes.get(attribute)
+    assert value == expected, "Expected %s '%s', but was '%s' for entity '%s'" % (
+        attribute,
+        expected,
+        value,
+        state.entity_id,
+    )
+
+
+async def async_call_service(
+    hass: HomeAssistant,
+    data: ComponentData,
+    device_id: int,
+    platform: str,
+    service: str,
+    extra_attribs: dict = None,
+) -> None:
+    """Call an HA service for a vera device."""
+    await hass.services.async_call(
+        platform,
+        service,
+        {
+            **{"entity_id": get_entity_id(get_device(device_id, data), platform)},
+            **(extra_attribs or {}),
+        },
+    )
+
+    await hass.async_block_till_done()
+
+
+async def async_configure_component(
+    hass: HomeAssistant,
+    requests_mocker: requests_mock.Mocker,
+    response_sdata: dict,
+    response_status: dict,
+    respone_lu_sdata: dict,
+) -> ComponentData:
+    """Configure the component with specific mock data."""
+    controller_url = "http://127.0.0.1:123"
+
+    component_data = ComponentData(
+        sdata=deepcopy(response_sdata),
+        status=deepcopy(response_status),
+        lu_sdata=deepcopy(respone_lu_sdata),
+    )
+
+    requests_mocker.get(
```

Comment:
Instead of using the requests mocker we should patch the library as necessary to avoid I/O in the tests.

Reply:
I've not been able to get patch to override `pyvera.VeraController.data_request`. Mocking request endpoints has been the most reliable solution.

Reply:
Try to patch and spec VeraController and the VeraDevice classes and set the return value of needed attributes.

Here's an example how to spec a class:
https://github.com/home-assistant/home-assistant/blob/206547a5d8abbcf7bc0463d312cc9f68cdcfce07/tests/components/homematicip_cloud/conftest.py#L130-L142

We should then have control of these instances.

Make sure to patch `init_controller` in the pyvera library as we use that to get a controller instance:
https://github.com/pavoni/pyvera/blob/4e9cf771627410a75bf12f0e7fb5250a68ecf910/pyvera/__init__.py#L60

We only care about the interfaces between pyvera and home assistant. We should not leak responsibility for the device specific interface like I/O and data parsing into home assistant in the tests.

I can try to give more suggestions if I get some more info on the specific problem with the patching.